File: 08 CP Priority Queue/Heap Ordering 2/d57_q2_heap_order2.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class d57_q2_heap_order2
{
    public static void main(String[] args)
    {
        Scanner Scanner = new Scanner(System.in);
        ArrayList<Student> students = new ArrayList<Student>();
        int N = Scanner.nextInt();
        int K = Scanner.nextInt();
        for(int i = 0; i < N; i++)
        {
            int score = 0;
            String name = Scanner.next();
            int n = Scanner.nextInt();
            for(int j = 0; j < n; j++)
            {
                int temp = Scanner.nextInt();
                score += temp;
            }
            students.add(new Student(name, score));
        }
        Collections.sort(students);
        Scanner.close();
        for(int i = 0; i < K; i++)
        {
            System.out.println(students.get(i));
        }
    }
}

class Student implements Comparable<Student>
{
    private int score;
    private String name;
    public Student(String name, int score)
    {
        this.name = name;
        this.score = score;
    }
    public int compareTo(Student other)
    {
        if(this.score > other.score) return -1;
        if(this.score < other.score) return 1;
        return this.name.compareTo(other.name);
    }
    public String toString()
    {
        return this.name + " " + this.score;
    }
}

File: 08 CP Priority Queue/Heap Ordering 2/d57_q2_heap_order2.cpp
#include<iostream>
#include<queue>
#include<algorithm>
using namespace std;

class Student
{
    public:
        // Class Properties
        int score;
        string name;
        // Constructor methods
        Student(string name, int score) : name(name), score(score) {}
        // Operator '<' Overloading
        // - Priority queue sorted by score in descending order
        // - Priority queue sorted by name in ascending order 
        bool operator<(const Student &other) const
        {
            if(this->score == other.score) return this->name > other.name;
            return this->score < other.score;
        }
};


int main()
{
    // This makes std::cin and std::cout faster
    ios_base::sync_with_stdio(false); cin.tie(0);

    // Input number of students (N) and number of output (K)
    int N,K;
    cin >> N >> K;
    priority_queue<Student> students;
    // Input student details
    while(N--)
    {
        string name;
        int n, score = 0;
        // Input name
        cin >> name >> n;
        // Find a sum of a score
        while(n--)
        {
            int temp;
            cin >> temp;
            score += temp;
        }
        // Create an object
        students.push(Student(name, score));
    }
    // Output
    while(K--)
    {
        cout << students.top().name << " " << students.top().score << endl;
        students.pop();
    }
    return 0;
}